### DISCUSSION TITLE
ATSAS 2.4 on CentOS / RHEL

### DISCUSSION THREAD

**Discussion by franke-biosaxs** [biosaxs-com/atsas-community#493] [ATSAS package in general]
On 2010-11-09 20:49:02, user jcbollinger wrote:

I was excited to spot ATSAS 2.4.0 packages available on the download site, complete with the promised Linux-native version of PRIMUS.  My hopes were dashed, however, when I discovered that the proffered RPM is incompatible with even the most recent updates to CentOS and RedHat Enterprise Linux.

There may be no escaping the new release's dependency on QT 4.5, but I can probably deal with that.  It will be very difficult, however, to work around the package's dependencies on more recent versions of the C and C++ standard libraries than CentOS and RHEL currently provide.

The ATSAS 2.3.2 RPM did not suffer from these problems.  Is there any chance of an RPM or tarball being provided that is built in the same development environment (including library versions) that that RPM was?

Thanks

**Comment by franke-biosaxs** [comment]
On 2010-11-10 10:25:30, user franke wrote:


jcbollinger wrote
> The ATSAS 2.3.2 RPM did not suffer from these problems.  Is there any chance of an RPM or tarball being provided that is built in the same development environment (including library versions) that that RPM was?

The previous release was linked all statically, in particular compiler support libraries and such. In other words, it brought everything needed.

I take it that the command line tools as dammin, sasref, gnom etc work without problems, but only primus does not?

The new release depends on Qt, btw the 4.5 release is almost 2 years old, current is 4.7, hence we can not feasible provide statically built binaries. To clarify, it would technically be possible if (1) we built static versions of Qt and (2) statically link the whole thing and (3) you don't mind a 200+MB primus executable that contains anything your system should already have anyway.

Instead, we could try to build our stuff on these somewhat outdated distros ourselves and provide the packages. Unfortunately, this does not easily work as we also rely on newer packages for building, e.g. many bugs in the RHEL system compiler (gcc/gfortran-4.1, iirc) have been fixed in later versions because ATSAS wouldn't compile properly. Hence we filed bug reports and patches; new versions, 4.3 and onward (current is 4.5!), work fine for us.

To summarize: if you want recent software and to use the latest developments, you need to use a fairly recent distribution as well.

HTH

Daniel

**Comment by franke-biosaxs** [comment]
On 2010-11-10 16:28:54, user jcbollinger wrote:

My apologies if this appears twice.  A posting error the first time around left me uncertain whether the original version was accepted, but it did not appear to have been.

franke wrote
> 
> jcbollinger wrote
> > The ATSAS 2.3.2 RPM did not suffer from these problems.  Is there any chance of an RPM or tarball being provided that is built in the same development environment (including library versions) that that RPM was?
> 
> The previous release was linked all statically, in particular compiler support libraries and such. In other words, it brought everything needed.
> 
> I take it that the command line tools as dammin, sasref, gnom etc work without problems, but only primus does not?

Thanks for your swift response.  No, few, if any, of the command-line tools work on fully-updated CentOS or RHEL 5.5.  As I said, the tools rely on a more recent version of the standard C library than CentOS or RHEL currently provide.  A few likewise rely on a more recent version of the standard C++ library.  Here's an excerpt from the ldd output:
> ./dammin: /lib/libc.so.6: version 'GLIBC_2.11' not found (required by ./dammin)

Please note that the linker is not complaining about a missing *library* ; rather, it is complaining that the library it finds lacks the 'GLIBC_2.11' feature.  That cannot be solved by installing a forwards-compatibility library in parallel with the standard one, as the two use the same file name.  Upgrading glibc and libstdc++ is not feasible, as too many other packages depend on exact versions of those libraries.

franke wrote
> The new release depends on Qt, btw the 4.5 release is almost 2 years old, current is 4.7, hence we can not feasible provide statically built binaries. To clarify, it would technically be possible if (1) we built static versions of Qt and (2) statically link the whole thing and (3) you don't mind a 200+MB primus executable that contains anything your system should already have anyway.
> 
> Instead, we could try to build our stuff on these somewhat outdated distros ourselves and provide the packages. Unfortunately, this does not easily work as we also rely on newer packages for building, e.g. many bugs in the RHEL system compiler (gcc/gfortran-4.1, iirc) have been fixed in later versions because ATSAS wouldn't compile properly. Hence we filed bug reports and patches; new versions, 4.3 and onward (current is 4.5!), work fine for us.
> 
> To summarize: if you want recent software and to use the latest developments, you need to use a fairly recent distribution as well.

RHEL 5.5 and CentOS 5.5 are less than 8 months old and actively maintained.  For strategic reasons they choose to be careful about adopting the latest developments.  CentOS, at least, seems fairly popular in the research community, so I doubt I will be the only person to encounter this difficulty.

If Qt is the main obstacle to static linkage, then would you perhaps consider a release in which only Qt is dynamically linked?

**Comment by franke-biosaxs** [comment]
On 2010-11-10 17:00:56, user franke wrote:


jcbollinger wrote
> RHEL 5.5 and CentOS 5.5 are less than 8 months old and actively maintained.  For strategic reasons they choose to be careful about adopting the latest developments.  CentOS, at least, seems fairly popular in the research community, so I doubt I will be the only person to encounter this difficulty.

For the sake of the argument, I just checked which versions of vital packages are shipped with RHEL and CentOS: cmake-2.6, latest is 2.8; gcc-4.1, current is 4.5; Qt-4.4, latest is 4.7. It may be a recent distrubtion, nonetheless, the packages shipped are somewhat dated.

To make a long story short, from a technical point of view, yes it is possible to spend a week or more to adjust our repository to deal with the older packages. I can backport features of cmake and the compiler that are available in later releases to the packages provided by RHEL/CentOS. If that would improve their stability, the reason they were chosen over later releases, is debatable. If there is time, I shall see if I'm overly pessimistic.

To make at least the command-line applications in the package work, I can try to find and eliminate the dependency on ```
$> nm -u dammin | grep GLIBC_2.11
U __longjmp_chk@@GLIBC_2.11
```
The next recent requirement would be GLIBC_2.3.4 which should be available. I wasn't aware of this, will see what I can do.

jcbollinger wrote
> If Qt is the main obstacle to static linkage, then would you perhaps consider a release in which only Qt is dynamically linked?

I considered to build two packages, atsas and atsas-gui, but given the amount of time I can allocate for this kind of work (hardly any), I decided not to.

**Comment by franke-biosaxs** [comment]
On 2010-11-10 17:39:44, user jcbollinger wrote:


franke wrote
> To make a long story short, from a technical point of view, yes it is possible to spend a week or more to adjust our repository to deal with the older packages. I can backport features of cmake and the compiler that are available in later releases to the packages provided by RHEL/CentOS. If that would improve their stability, the reason they were chosen over later releases, is debatable. If there is time, I shall see if I'm overly pessimistic.
> 
> To make at least the command-line applications in the package work, I can try to find and eliminate the dependency on```
> $> nm -u dammin | grep GLIBC_2.11
> U __longjmp_chk@@GLIBC_2.11
> ```
> The next recent requirement would be GLIBC_2.3.4 which should be available. I wasn't aware of this, will see what I can do.

Thank you, I appreciate your efforts.

franke wrote
> 
> jcbollinger wrote
> > If Qt is the main obstacle to static linkage, then would you perhaps consider a release in which only Qt is dynamically linked?
> 
> I considered to build two packages, atsas and atsas-gui, but given the amount of time I can allocate for this kind of work (hardly any), I decided not to.

Given that your time is limited, is there any possibility of a source distribution, so that I could work on the problem myself?  I would be willing agree to reasonable non-disclosure and non-redistribution terms, and to give you back the result if I am successful.

**Comment by franke-biosaxs** [comment]
On 2010-11-10 18:15:57, user franke wrote:


franke wrote
> To make at least the command-line applications in the package work, I can try to find and eliminate the dependency on```
> $> nm -u dammin | grep GLIBC_2.11
> U __longjmp_chk@@GLIBC_2.11
> ```
> The next recent requirement would be GLIBC_2.3.4 which should be available. I wasn't aware of this, will see what I can do.

The package was compiled on Ubunutu-10.04. FIrst testing indicates that it may be a Ubuntu-specific modification to the system compiler libraries. The compiler libraries compiled from vanilla sources do not seem to depend on this. Great. Investigation continues.

jcbollinger wrote
> Given that your time is limited, is there any possibility of a source distribution, so that I could work on the problem myself?  I would be willing agree to reasonable non-disclosure and non-redistribution terms, and to give you back the result if I am successful.

Thanks for the offer, but as far as I know, no chance. I brought this up myself a couple of times, but the move was rejected.

**Comment by franke-biosaxs** [comment]
On 2011-03-16 17:46:32, user bene wrote:


franke wrote
> To make at least the command-line applications in the package work, I can try to find and eliminate the dependency on```
> $> nm -u dammin | grep GLIBC_2.11
> U __longjmp_chk@@GLIBC_2.11
> ```
> The next recent requirement would be GLIBC_2.3.4 which should be available. I wasn't aware of this, will see what I can do.

Hello franke,

Have you been able to make any progress on this?  Red Hat is a de facto standard at many research institutions in the US, and the fact that the ATSAS software won't run on Red Hat 5 systems is going to limit its adoption.

**Comment by franke-biosaxs** [comment]
On 2011-03-17 16:19:25, user bene wrote:

And not to belabor the point too much, but you asked what linux distributions people were using about 2.5 years ago:

[]()

And the two external people that replied both are using Red Hat/CentOS.

Scientific Linux from Fermilab/CERN is also based on the Red Hat sources.

I'll be quiet now.

**Comment by franke-biosaxs** [comment]
On 2011-04-13 13:00:38, user lievbuts wrote:

For me it would also be useful to run certain ATSAS programs on systems with older glibc versions. We have just collected a large amount of SAXS data, and would like to run the calculations on a cluster system. At the moment we have access to the university cluster, which is running a very old Red Hat derivative with glibc version 2.3.4. However, next week we will be installing a brand new cluster of our own, which will come with Scientific Linux 5.6. Since this is based on Red Hat 5.6, it will apparently have glibc version 2.5.

**Comment by franke-biosaxs** [comment]
On 2011-04-13 13:24:39, user franke wrote:

We are preparing the release of atsas-2.4.1; for this I plan natively built packages for
  - CentOS 5.5 (32/64 bit)
Scientific Linux 5 (64bit only)
Debian 5 (64bit only)
[K]Ubuntu 10.4 (32/64 bit) This should resolve most problems.
Stay tuned, the release will be announced on this board :)

**Comment by franke-biosaxs** [comment]
On 2011-04-13 15:00:43, user lievbuts wrote:


franke wrote
> We are preparing the release of atsas-2.4.1; for this I plan natively built packages for
>   - CentOS 5.5 (32/64 bit)
> Scientific Linux 5 (64bit only)
> Debian 5 (64bit only)
> [K]Ubuntu 10.4 (32/64 bit)This should resolve most problems.
> Stay tuned, the release will be announced on this board

Excellent news! Thank you.
